DOC-11139 explain function (#127)
couchbaselabs/docs-devex: modules/n1ql/pages/n1ql-language-reference/explainfunction.adoc

* If an external function defines an alias for an N1QL() call, EXPLAIN FUNCTION cannot return its line number.

== Prerequisites

To execute EXPLAIN FUNCTION on an inline user-defined function, you must have the necessary execute function permission on that function. 

* For a global internal function, you must have the Execute Global Functions role. 

* For a scoped internal function, you must have the Execute Scope Functions role, with permissions on the specified bucket and scope.

To execute EXPLAIN FUNCTION on an external user-defined JavaScript function, you must have the necessary execute external function permission on that function. 

* For a global JavaScript function, you must have the Execute Global External Functions role. 

* For a scoped JavaScript function, you must have the Execute External Scope Functions role, with permissions on the specified bucket and scope.

You must also have the necessary privileges required for the {sqlpp} statements inside the function.

For more information about user roles, see
xref:learn:security/authorization-overview.adoc[Authorization].

== Syntax

[source,ebnf]
----

> include::partial$grammar/utility.ebnf[tag=explainfunction]

Reviewer: Correct tag name for EBNF syntax

```suggestion
include::partial$grammar/utility.ebnf[tag=explain-function]
```

----

> image::n1ql-language-reference/explainfunction.png["Syntax diagram: refer to source code listing", align=left]

Reviewer: Correct file name for diagram

```suggestion
image::n1ql-language-reference/explain-function.png["Syntax diagram: refer to source code listing", align=left]
```

> The statement consists of the `EXPLAIN FUNCTION` keywords followed by the name of the user-defined function that has the execution plan you want to review.

Reviewer: Is it worth adding more details about the function name here? Does the function name in EXECUTE FUNCTION have the same structure and constraints as the function name in other function commands such as DROP FUNCTION? I'm guessing that it does, but please check.

The following text is adapted from the DELETE FUNCTION page.

```suggestion
function:: (Required) Refer to <<name>> below.

[[name]]
=== Function Name

[source,ebnf]
----
include::partial$grammar/ddl.ebnf[tag=function]
----

image::n1ql-language-reference/function.png["Syntax diagram: refer to source code listing", align=left]

The name of the function.
This is usually an unqualified identifier, such as `func1` or `{backtick}func-1{backtick}`.
In this case, the path to the function is determined by the current xref:n1ql:n1ql-intro/queriesandresults.adoc#query-context[query context].

To get the plan for a global function in a particular namespace, the function name must be a qualified identifier with a namespace, such as `default:func1`.
Similarly, to get the plan for a scoped function in a particular scope, the function name must be a qualified identifier with the full path to a scope, such as `default:{backtick}travel-sample{backtick}.inventory.func1`.
Refer to xref:n1ql-language-reference/createfunction.adoc#context[Global Functions and Scoped Functions] for more information.

NOTE: The name of a user-defined function _is_ case-sensitive, unlike that of a built-in function.
You must get the plan for the user-defined function using the same case that was used when it was created.
```

Author: Confirmed, will add an xref to the CREATE FUNCTION description of function name

== Examples

> === Inline Function Example
> 
> In this example, you create an inline function named `func1` and then request the plan for its subquery.
> 
> ====

Reviewer: We usually use titles for the examples in the references rather than individual headings. Also suggest putting the "in this example" intro paragraph into the actual example.

```suggestion
.Inline Function Example

====
In this example, you create an inline function named `func1` and then request the plan for the embedded statement.

```

[source,sqlpp]
----
include::n1ql:example$utility/explainfunction.n1ql[]
----

.Results
[source,json]
----
include::n1ql:example$utility/explainfunction.jsonc[]
----
====

> === External Function Example
> 
> This example assumes that you have defined a JavaScript library named  `lib1`.
> 
> You then add a JavaScript function named `function1` to that library as follows:

Reviewer: Example title, add example markup

```suggestion
.External Function Example
====
This example assumes that you have defined a JavaScript library named  `lib1`.

You then add a JavaScript function named `function1` to that library as follows:
```

[source,javascript]
----
include::n1ql:example$utility/explainfunctionjs.js[]
----

<.> An embedded {sqlpp} statement.

> <.> An N1QL() call that executes a {sqlpp} statement.

Reviewer: Very minor point: I think most users would pronounce it "nickel" (because we tell them to), so the indefinite article would be "a". Same throughout.

```suggestion
<.> A N1QL() call that executes a {sqlpp} statement.
```

You then create the corresponding inline user-defined function for that JavaScript function, named `jsfunction1`, and request the plan for its embedded {sqlpp} expression:

[source,SQL++]
----
include::n1ql:example$utility/explainfunctionjs.n1ql[]
----

.Results
[source,json]
----
include::n1ql:example$utility/explainfunctionjs.jsonc[]
----

<.> The line number in the JavaScript function that includes an N1QL() call. 

> <.> The query plan for the embedded {sqlpp} statement.

Reviewer: Close example markup

```suggestion
<.> The query plan for the embedded {sqlpp} statement.
====
```

= EXPLAIN FUNCTION

> :description: EXPLAIN FUNCTION provides information about the execution plan for the {sqlpp} subqueries or embedded statements defined for a specified user-defined function. 

Reviewer 2: Could we modify the introduction to imply that this returns query plans for SQL++ within/ inside the user-defined function definition ?

Author: @dhanyagowrish I thought that "subqueries" and "embedded statements" covered this idea, would you be able to rephrase more accurately or so that it is easier to understand?

Here's my attempt to put the focus more on UDF, let me know if this is ok: 

For a specified user-defined function, you use EXPLAIN FUNCTION to expose the execution plan for the {sqlpp} subqueries or embedded statements that it contains.

Reviewer 2: What I meant was that, it would be ideal to imply that EXPLAIN FUNCTION returns query plans for statements ( whether subqueries or embedded statements ) within/ inside the UDF definition.

Author: @simon-dew can you suggest phrasing for the description to accommodate Dhanya's request?

Reviewer: I would have said "defined _within_ a specified user-defined function", but your new phrasing works just as well

:page-topic-type: reference
:imagesdir: ../../assets/images

// TEMP
include::partial$n1ql-language-reference/collapsible-style.adoc[]

[abstract]
{description}

You can supply the name of an inline or external user-defined function.

* For an inline function, EXPLAIN FUNCTION returns the query plans for all of the subqueries present in the function body.
+
For more information about inline functions, see xref:n1ql-language-reference/userfun.adoc[].

* For an external function, EXPLAIN FUNCTION returns the query plans for all embedded {sqlpp} queries inside the referenced JavaScript body, or the line number on which a N1QL() call appears.
+
For more information about user-defined functions with JavaScript, see xref:guides:javascript-udfs.adoc[].

The following constraints apply:

* If a user-defined function itself contains other, nested user-defined function executions, EXPLAIN FUNCTION generates the query plan for the specified function only, and not for its nested {sqlpp} queries.

* If an external function defines an alias for a N1QL() call, EXPLAIN FUNCTION cannot return its line number.

> 

Reviewer 2: Please add this about how the value of line numbers is calculated - "The line numbers are calculated from the beginning of the JavaScript function definition."

* For a global internal function, you must have the Execute Global Functions role. 

* For a scoped internal function, you must have the Execute Scope Functions role, with permissions on the specified bucket and scope.

To execute EXPLAIN FUNCTION on an external user-defined JavaScript function, you must have the necessary execute external function permission on that function. 

* For a global JavaScript function, you must have the Execute Global External Functions role. 

* For a scoped JavaScript function, you must have the Execute External Scope Functions role, with permissions on the specified bucket and scope.

You must also have the necessary privileges required for the {sqlpp} statements inside the function.

For more information about user roles, see
xref:learn:security/authorization-overview.adoc[Authorization].

== Syntax

[source,ebnf]
----
include::partial$grammar/utility.ebnf[tag=explain-function]
----

image::n1ql-language-reference/explain-function.png["Syntax diagram: refer to source code listing", align=left]

The statement consists of the `EXPLAIN FUNCTION` keywords followed by the name of the user-defined function that has the execution plan you want to review.

> For more information about function name syntax and requirements, see xref:n1ql-language-reference/createfunction.adoc#syntax[CREATE FUNCTION].

Reviewer 2: Should the link for the syntax, etc for function name refer to the actual function name section? - i.e
 https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/execfunction.html#name ?

or 

https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/createfunction.html#inline-name

Author: @dhanyagowrish agreed, however there is some funky asciidoc markdown in there that prevented me from getting directly there. I'll set up time with Simon to see if I can get closer at least.

== Examples

.Inline Function Example

====
In this example, you create an inline function named `func1` and then request the plan for its subquery.

[source,sqlpp]
----
include::n1ql:example$utility/explainfunction.n1ql[]
----

.Results
[source,json]
----
include::n1ql:example$utility/explainfunction.jsonc[]
----
====

.External Function Example

====
This example assumes that you have defined a JavaScript library named  `lib1`.

You then add a JavaScript function named `function1` to that library as follows:

[source,javascript]
----
include::n1ql:example$utility/explainfunctionjs.js[]
----

<.> An embedded {sqlpp} statement.
<.> A N1QL() call that executes a {sqlpp} statement.

> You then create the corresponding inline user-defined function for that JavaScript function, named `jsfunction1`, and request the plan for its embedded {sqlpp} expression:

Reviewer 2: Could this line be modified?
	1	Modify: "....corresponding inline user-defined function..." to "… corresponding SQL++ user-defined function…"
( Since "inline udf" has a different meaning in sql++)
	2	Modify: "....request the plan for its embedded SQL++ expression..." -> to something alone the lines of "...request the plan information for the statements within the function definition.."
Since the output of EXPLAIN FUNCTION is not just plan info for embedded expressions

Author: @dhanyagowrish updated. I think I understand that you are trying to be inclusive of a line number being returned  as "plan information" though that isn't how I think of it...

Reviewer 2: Yes, that was my intention.
